torch/csrc/StorageSharing.cpp

```diff
@@ -456,98 +446,31 @@ static PyObject* THPStorage_newSharedCuda(PyObject* _unused, PyObject* args) {
       (size_t)THPUtils_unpackLong(_size_bytes) / sizeof(uint8_t);
   ptrdiff_t storage_offset_bytes =
       (ptrdiff_t)THPUtils_unpackLong(_offset_bytes);
-
-  const auto device = c10::checked_convert<c10::DeviceIndex>(
+  const auto device_index = c10::checked_convert<c10::DeviceIndex>(
       THPUtils_unpackLong(_device), "c10::DeviceIndex");
-  at::cuda::CUDAGuard device_guard(device);
-
-  if (PyObject_IsTrue(_event_sync_required)) {
-    // Ensure that producer prepared all tensor's data
-    std::string s_ipc_event_handle =
-        THPStorage_bytesAsHandleString(_event_handle);
-    if (s_ipc_event_handle.empty()) {
-      return nullptr;
-    }
-    auto ipc_event_handle = reinterpret_cast<const cudaIpcEventHandle_t*>(
-        s_ipc_event_handle.c_str());
-    cudaEvent_t event = nullptr;
-    cudaIpcOpenEventHandle(&event, *ipc_event_handle);
-    C10_CUDA_CHECK(
-        cudaStreamWaitEvent(c10::cuda::getCurrentCUDAStream(device), event, 0));
-  }
-
-  std::string s_handle = THPStorage_bytesAsHandleString(_handle);
-  if (s_handle.empty()) {
+  auto device = at::Device(device_type, device_index);
+  at::DeviceGuard device_guard(device);
+  std::string s_ipc_event_handle =
+      THPStorage_bytesAsHandleString(_event_handle);
+  bool event_sync_required = PyObject_IsTrue(_event_sync_required);
+  if (event_sync_required && s_ipc_event_handle.empty())
     return nullptr;
```

Comment:
This doesn't look exactly semantics preserving. In the old code, even if the ipc event handle was empty, if an event sync was required, we would trigger a cuda stream wait event. In the new code, you always skip the sync.